### DISCUSSION TITLE
Leveraging BMW Browser and Carplay webserver solutions (proven for Tesla)

### DISCUSSION THREAD

**Discussion by m0ntev** [BimmerGestalt/AAIdrive#741] [Ideas]
Hi,
For the Tesla, people have built solutions to enable carplay by having a device (e.g. raspberry pi) be a web-server that presents carplay for the tesla browser to display. BMW does have a browser that can be enabled by coding.

Would it be possible to do something similar e.g.
- Use android device as carplay web server
- Use the AAI drive app / MyBMW to have an app on iDrive
- Use the BMW built in browser

See more info:
https://github.com/marcraft2/tesla-carplay (web server project 1)
https://x.com/mikegapinski?s=20 (web server project 2)
https://carlinkitcarplay.com/products/carlinkit-t2c-tesla-wireless-apple-carplay-adapter (web server project 3)

https://www.bimmerfest.com/threads/updating-hu_nbt-activate-bmw-online-and-internet-via-tool32.735962/page-3#post-8179599 (BMW Browser activation (NBT))
Browser in action, controlled by iDrive: https://www.youtube.com/watch?v=y53m24CtITU

**Comment by hufman** [comment]
That's very interesting!
Does that work for NBT Evo too? NBT Evo also supports regular Carplay directly, what functionality does the Carplay web interface provide on an Android?
What role do you propose AAIdrive provide? How would the car connect to a web server on the phone? AAIdrive (actually MyBMW) talks to the car over USB Serial or Bluetooth Serial, and doesn't have full HTTP access.

**Reply by m0ntev** [reply]
Hi Walter
My guess would be that AAIdrive provides the hosting of a carplay server on the android device, and a substitute connection to the NBT in lieu of it not having wifi to connect to the browser. Not sure though!  I expect the NBT browser is to crummy.

**Reply by hufman** [reply]
As far as I know, there is no way for the car to open TCP connections to the phone without Wifi. AAIdrive would not provide any benefit, a separate project would be better positioned to implement this idea.

**Reply by mwarrenus** [reply]
FWIW, the CarlinKit T2C product uses Bluetooth PAN-NAP (Personal Area Network - Network Access Point) in place of WiFi to connect to the phone hotspot for internet access. The BT connection shows up as "Ethernet" in the iPhone settings. The T2C also creates a WiFi network that you connect to the Tesla head unit for the T2C to display CarPlay in Tesla's web browser. This WiFi connection also causes all Internet communication from the Tesla to go through the T2C, over the BT PAN, and then thru phone.

**Reply by hufman** [reply]
That is very interesting!
I have no idea how to set up Bluetooth PAN on a BMW, my Mini's NBT Evo probably doesn't even support it since they went with LTE.
I'd be curious to hear your experiments though.
